generator/internal/http_annotation_parser.cc

```diff
@@ -0,0 +1,281 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "generator/internal/http_annotation_parser.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/log.h"
+#include "absl/strings/match.h"
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+namespace {
+
+template <typename T>
+struct ParseSuccess {
+  T value;
+  std::size_t end;  // where parsing finished
+};
+template <typename T>
+ParseSuccess<T> MakeParseSuccess(T value, std::size_t end) {
+  return ParseSuccess<T>{std::move(value), end};
+}
+
+template <typename T>
+using ParseResult = StatusOr<ParseSuccess<T>>;
+
+Status ParseError(absl::string_view input, std::size_t offset,
+                  absl::string_view expected,
+                  internal::ErrorInfoBuilder builder) {
+  return internal::InvalidArgumentError(
+      absl::StrCat("error parsing path template, expected", expected,
+                   " at offset ", offset, "\n", input, "\n",
+                   std::string(offset, ' '), "^"),
+      std::move(builder));
+}
+
+ParseResult<PathTemplate::Segments> ParseSegments(absl::string_view input,
+                                                  std::size_t offset);
+
+ParseResult<PathTemplate::Segments> ParsePlainSegments(absl::string_view input,
+                                                       std::size_t offset);
+
+ParseResult<PathTemplate::Segment> ParsePlainSegment(absl::string_view input,
+                                                     std::size_t offset);
+
+ParseResult<PathTemplate::Segment> ParseSegment(absl::string_view input,
+                                                std::size_t offset);
+
+ParseResult<PathTemplate::Segment> ParseVariable(absl::string_view input,
+                                                 std::size_t offset);
+
+ParseResult<PathTemplate::Segments> ParsePlainSegments(absl::string_view input,
+                                                       std::size_t offset) {
+  PathTemplate::Segments segments;
+  while (input.size() != offset) {
+    auto s = ParsePlainSegment(input, offset);
+    if (!s) return std::move(s).status();
+    segments.push_back(
+        std::make_shared<PathTemplate::Segment>(std::move(s->value)));
+    offset = s->end;
+    if (input.size() == offset || input[offset] != '/') break;
+    ++offset;
+  }
+  if (segments.empty()) {
+    return ParseError(input, offset, " segment", GCP_ERROR_INFO());
+  }
+  return MakeParseSuccess(std::move(segments), offset);
+}
+
+ParseResult<PathTemplate::Segments> ParseSegments(absl::string_view input,
+                                                  std::size_t offset) {
+  PathTemplate::Segments segments;
+  while (input.size() != offset) {
+    auto s = ParseSegment(input, offset);
```

Comment:
Q.  Is it worth using a parameter to distinguish `ParsePlainSegment` from `ParseSegment` rather than using separate functions, despite the EBNF?  I'm not sure.

Reply:
I am not sure either.  I refactored the code in a way that (I think) still makes it easy to navigate the parsing functions.